### PR TITLE
Limit group by to MaxGroupByPoints (currently 100,000)

### LINKED ISSUE #1992: Setting a LIMIT to 2147483647 causes server to crash
Need to investigate this. I think this is the case, reported by a user. If the limit overflows we should just set it to whatever maxInt is.

### PR DESCRIPTION
**Pull request by corylanou**
Fixes #1992

Also fixes a bug in `results.Error()`.  It wasn't looking into the slice of series to see if those also had an error.  I needed to fix that to make my tests work otherwise I couldn't see the error.

### CODE REVIEW

**Comment by pauldix** [comment]
One possible enhancement we could make to this is to either use the limit or the end time of the query.

File: server_test.go [hunk 1]
```diff
@@ -1818,6 +1818,30 @@ func TestServer_ExecuteWildcardQuery(t *testing.T) {
...
+	s := OpenServer(c)
+	defer s.Close()
+	s.CreateDatabase("foo")
+	s.CreateRetentionPolicy("foo", &influxdb.RetentionPolicy{Name: "raw", Duration: 1 * time.Hour})
+	s.SetDefaultRetentionPolicy("foo", "raw")
+	s.CreateUser("susy", "pass", false)
+
+	// Write series with one point to the database.
+	s.MustWriteSeries("foo", "raw", []influxdb.Point{{Name: "cpu", Tags: map[string]string{"region": "us-east"}, Timestamp: mustParseTime("2000-01-01T00:00:00Z"), Fields: map[string]interface{}{"value": float64(10)}}})
+	s.MustWriteSeries("foo", "raw", []influxdb.Point{{Name: "cpu", Tags: map[string]string{"region": "us-east"}, Timestamp: mustParseTime("2000-01-01T00:00:10Z"), Fields: map[string]interface{}{"value": 20}}})
```
Comment (otoolep) [inline server_test.go]: Doesn't this case only require 1 write? If so, the code could be shorter.

File: server_test.go [hunk 2]
```diff
@@ -1818,6 +1818,30 @@ func TestServer_ExecuteWildcardQuery(t *testing.T) {
...
+	s.MustWriteSeries("foo", "raw", []influxdb.Point{{Name: "cpu", Tags: map[string]string{"region": "us-west"}, Timestamp: mustParseTime("2000-01-01T00:00:20Z"), Fields: map[string]interface{}{"value": 30}}})
+
+	// GROUP BY * (wildcard) with fill(0) and large limit.
+	results := s.ExecuteQuery(MustParseQuery(`select mean(value)  from cpu where  region='us-west'  and time > '2000-01-01T00:00:00Z' group by time(300s), * fill(0)  limit 2147483647 slimit 1`), "foo", nil)
+	if results.Error() == nil {
+		t.Fatal("expected error: received nil")
```
Comment (otoolep) [inline server_test.go]: This failure message could be improved. It could at least include the name of the test.

File: server_test.go [hunk 3]
```diff
@@ -1818,6 +1818,30 @@ func TestServer_ExecuteWildcardQuery(t *testing.T) {
...
+// that results in a grouping of over 100,000 points returns an error
+func TestServer_ExecuteGroupByFillLimit(t *testing.T) {
+	c := test.NewMessagingClient()
+	defer c.Close()
+	s := OpenServer(c)
+	defer s.Close()
+	s.CreateDatabase("foo")
+	s.CreateRetentionPolicy("foo", &influxdb.RetentionPolicy{Name: "raw", Duration: 1 * time.Hour})
+	s.SetDefaultRetentionPolicy("foo", "raw")
+	s.CreateUser("susy", "pass", false)
```
Comment (otoolep) [inline server_test.go]: No need to create "susy".

File: server_test.go (hunk 2 above)
Comment (corylanou) [inline server_test.go]: If it fails, go already puts out the name of the test like this:

```sh
--- FAIL: TestServer_ExecuteGroupByFillLimit (0.15s)
  server_test.go:1841: expected error: received nil
```

Comment (otoolep) [inline server_test.go]: Ah, OK, good enough -- thanks.

File: server_test.go [hunk 4]
```diff
@@ -1506,6 +1506,37 @@ func TestServer_LimitAndOffset(t *testing.T) {
 	}
 }
 
+// Ensure the server can execute a group by with a fill(0) and very large limit
+// that results in a grouping of over 100,000 points returns an error
+func TestServer_ExecuteGroupByFillLimit(t *testing.T) {
```
Comment (pauldix) [inline server_test.go]: I think `server_integration_test.go` would be a better place for this. Easier to just add this to the table tests.

File: influxql/engine.go [hunk 5]
```diff
@@ -120,16 +111,26 @@ func (m *MapReduceJob) Execute(out chan *Row, filterEmptyResults bool) {
...
 				Tags: m.TagSet.Tags,
 			}
-
 			return
 		} else {
-			pointCountInResult = m.stmt.Limit
-			if m.stmt.Offset+m.stmt.Limit > pointCountInResult {
-				pointCountInResult = pointCountInResult - m.stmt.Offset
+			if m.stmt.Offset+m.stmt.Limit > m.stmt.Limit {
+				if m.stmt.Limit-m.stmt.Offset < pointCountInResult {
```
Comment (pauldix) [inline influxql/engine.go]: This shouldn't be inside this if statement, should it? This won't set the max limit to the lesser of either the point count or limit in cases where the offset is 0.

File: server_test.go (hunk 4 above)
Comment (corylanou) [inline server_test.go]: I was headed that direction and realized that those tests aren't set up to check for error conditions.  I do agree it should be there.  I'll have to do some refactoring of that test suite to make this one possible.

File: influxql/engine.go (hunk 5 above)
Comment (corylanou) [inline influxql/engine.go]: Yeah, this was not very clear to me.  I was looking for some comments or tests but could't find any to help determine some of the edge cases.  I'll write out some extended tests today to make sure we have all the edge cases covered, as well as add some documents in the code so future reviewers have a good grasp on what we are trying to accomplish.

File: influxql/engine.go [hunk 6]
```diff
@@ -138,7 +134,18 @@ func (m *MapReduceJob) Execute(out chan *Row, filterEmptyResults bool) {
...
-		t := startTimeBucket + (int64(i) * m.interval * int64(m.stmt.Offset+1))
+		var t int64
+		if m.stmt.Offset > 0 {
+			t = startTimeBucket + (int64(i+1) * m.interval * int64(m.stmt.Offset))
+		} else {
+			t = startTimeBucket + (int64(i+1) * m.interval) - m.interval
+		}
+
+		// If we start getting out of our max time range, then truncate values and return
+		if t > m.TMax && !isRaw {
```
Comment (pauldix) [inline influxql/engine.go]: Would this ever happen? Shouldn't the size of `resultTimes` be fixed to the exact number of buckets we want?

File: cmd/influxd/server_integration_test.go [hunk 7]
```diff
@@ -952,25 +961,23 @@ func runTestsData(t *testing.T, testName string, nodes Cluster, database, retent
 		},
 	}
 
+	// See if we should run a subset of this test
+	testPrefix := os.Getenv("TEST_PREFIX")
```
Comment (otoolep) [inline cmd/influxd/server_integration_test.go]: Can you give me an example of `TEST_PREFIX`?

Comment (otoolep) [inline cmd/influxd/server_integration_test.go]: `limit and` for example?

File: influxql/ast.go [hunk 8]
```diff
@@ -598,7 +598,7 @@ type SelectStatement struct {
 	groupByInterval time.Duration
 
 	// if it's a query for raw data values (i.e. not an aggregate)
-	RawQuery bool
+	IsRawQuery bool
```
Comment (otoolep) [inline influxql/ast.go]: is this Go style?

File: influxql/parser.go [hunk 9]
```diff
@@ -577,6 +577,15 @@ func (p *Parser) parseSelectStatement(tr targetRequirement) (*SelectStatement, e
 		return nil, err
 	}
 
+	// Set if the query is a raw data query or one with an aggregate
+	stmt.IsRawQuery = true
```
Comment (otoolep) [inline influxql/parser.go]: In our code we have this concept of "sub statements". Two "select" statements are kicked out from a statement like `SELECT value,mean(value) FROM....` and processing sometimes takes place on sub-statements. Will the `IsRawQuery` flag be set correctly on substatements? I have a feeling that this flag should be a method on the `SelectStatement` type, and *not* determined at parse-time.

cc @benbjohnson

Comment (pauldix) [inline influxql/parser.go]: `SELECT value, mean(value)` isn't a valid query. If you have an aggregate, you have a group by time, so it doesn't work with that.

The exception to this would be `SELECT host, min(value) from cpu`, which will be a valid query when it's wired up (#1577). However, that will still be an aggregate query, not a raw one.

File: influxql/ast.go (hunk 8 above)
Comment (pauldix) [inline influxql/ast.go]: Idk, is it? I realized that it was a bool and since it's not in a struct that contains nothing but bools, I figured it would be clearer to have the `Is` there. Curious to hear what others think. @benbjohnson, @dgnorton?

File: influxql/parser.go (hunk 9 above)
Comment (otoolep) [inline influxql/parser.go]: OK, but doesn't your second example still result in two substatements within the code? One for `host` and one for `min(value)`. It's been a while since I looked at the relevant code, I just recall that "substatement" concept and want to check that it's all OK.

Comment (pauldix) [inline influxql/parser.go]: No, the second example will be a single statement. The `min` aggregate
function will basically pass different things to the reducer based on the
other parts of the query. When the `min` mapper and reducer functions are
initialized, they'll take arguments for tags and fields that should be
included.

Those will then be processed by the processor that handles the output from
the reducer. It's a bit tricky, but the identification of the other fields
needs to happen within the query scan that handles the min.

Comment (otoolep) [inline influxql/parser.go]: In any event, would it more robust if `IsRawQuery` was a method, and not a flag? Yes, it means we might have to execute the method a couple of times, but it seems to me to be more coherent if `IsRawQuery` didn't require a SELECT statement to be brought into existence via parsing for `IsRawQuery` to be correct.

Comment (pauldix) [inline influxql/parser.go]: There isn't really a concept of a substatement anymore. Queries should be processed and planned as is. However, that did remind me that `IsRawQuery` needed to be included in the `Clone` method. I'd rather leave it as a property for now, but if you feel VERY strongly about it I'll probably fold and make it a method ;)

Comment (otoolep) [inline influxql/parser.go]: I think I might feel strongly about this. For example, look at this?

https://github.com/influxdb/influxdb/blob/master/influxql/ast.go#L756

Why is this then a method and not a property? My analogy may be poor, and if so I don't understand the code please let me know.

Comment (otoolep) [inline influxql/parser.go]: Ditto for `OnlyTimeDimensions()`.

Comment (pauldix) [inline influxql/parser.go]: Those should probably be properties that get set on parse. Like everything else here: https://github.com/influxdb/influxdb/blob/master/influxql/parser.go#L518-L580

And actually, `Aggregated` should be removed because it's representing the same thing as `IsRawQuery` except it's using different logic to do so. We'd be better off removing that and keeping things consistent.

Comment (otoolep) [inline influxql/parser.go]: OK. It was the inconsistency that threw me, but if we, in principle, want to move to properties set on parse, then it all makes sense.

File: influxql/ast.go (hunk 8 above)
Comment (dgnorton) [inline influxql/ast.go]: In this case, I prefer `IsRawQuery`.

File: cmd/influxd/server_integration_test.go (hunk 7 above)
Comment (dgnorton) [inline cmd/influxd/server_integration_test.go]: I like this idea.  What if we made it a regex, kinda like the `-run` flag?

Comment (corylanou) [inline cmd/influxd/server_integration_test.go]: I'm all for it.  I likely won't do it on this PR due to needing to get it shipped.  But if I don't get to it, feel free to add it as well.  Makes a lot of sense to work just like the run flag.

File: influxql/ast.go (hunk 8 above)
Comment (benbjohnson) [inline influxql/ast.go]: I prefer `IsRawQuery` too. `RawQuery` makes me think that it's a `Query` type if I saw it used somewhere.

**Comment by otoolep** [comment]
Made the test @corylanou requested, but now the tests are failing. Looking into it.

**Comment by otoolep** [comment]
I see the original implementation of `Aggregated` did a walk, so `IsRawQuery` is not an exact copy.

**Comment by pauldix** [comment]
@otoolep yeah, `IsRawQuery` should actually be doing the same thing `Aggregated()` is doing, but inverse :)

**Comment by otoolep** [comment]
D'oh @pauldix of course, reverse it.

**Comment by pauldix** [comment]
LGTM :ship:

**Comment by corylanou** [comment]
@otoolep Super big awesome thank you for picking this up!
